### DISCUSSION TITLE
Anything like x-unmount?

### DISCUSSION THREAD

**Discussion by mariusGundersen** [alpinejs/alpine#1217] [1. Help]
There is the `x-init` attribute, which is similar to `onMount`, but there doesn't seem to be anything equivalent to `onUnmount`, like `x-unmount`. Is there any way to get notified when a component is unmounted, so that it can clean up for example `setInterval` or other resources.

**Comment by KevinBatdorf** [comment]
How are you unmounting the component?

**Reply by mariusGundersen** [reply]
It's inside an x-for loop, and the list changes.

**Comment by mariusGundersen** [comment]
I guess the silence here means that there is no way to do this? That's very unfortunate. It should bea priority for the next version.

**Reply by KevinBatdorf** [reply]
Not with Alpine V2. You could possibly add your own watcher to manage cleanup. So you would watch the array of items you're looping over and keep track of what was removed, then run a cleanup function.

**Reply by sukei** [reply]
I'm interested by this too. I'm using Hotwire Turbo with AlpineJS and components are replaced by it. I would like to handle teardown (closing EventSource stream) the right way. No doubt you got such a mechanism to cleanup the event listeners in the framework. As of today, I bypassed this issue by creating a directive that provide such mechanism.

**Comment by 3xau1o** [comment]
Currently can be manually implemented using `MutationObserver` on the other hand `x-clean` would be nice adition

**Comment by ekwoka** [comment]
In the data component, the `destroy` method is called on cleanup. Basically the inverse of `init`

**Reply by bennadel** [reply]
Just thinking out-loud here, but you could basically create a directive for that like:

```js
// x-destroy="doSomething()"
Alpine.directive( "destroy", DestroyDirective );

function DestroyDirective( el, metadata, framework ) {

	framework.cleanup(
		() => {
			framework.evaluate( metadata.expression );
		}
	);

}
```

I think that would accomplish this.

**Reply by 3xau1o** [reply]
ref https://alpinejs.dev/globals/alpine-data#destroy-functions
